### PR TITLE
Update metadata validation and generation

### LINKED ISSUE #2560: [Discussion] image-based gadgets: How to mark the maps used by a gadget to generate information?
## Introduction

The gadgets generate information using different kind of maps: tracers use perf event arrays or ring buffers, snapshotters use ebpf iterators and toppers and profilers use hashmaps. Inspektor Gadget needs to know what are the maps (and the types) that are used by the gadget to generate the information in order to be able to run the gadget and collect the information it generates. 

Currently Inspektor Gadget supports two ways to set this information. Any of the two methods needs to be used. If both methods are used, Inspektor Gadget validates that them contain the same information.

### Metadata file 

The metadata file, besides other things, allows to set the kind of information that is provided by the gadget:

- https://github.com/inspektor-gadget/inspektor-gadget/blob/8c433c4b29399cabd6ac233933c602f76c3c7c5f/gadgets/trace_exec/gadget.yaml#L3-L6
- https://github.com/inspektor-gadget/inspektor-gadget/blob/79eb65512419557898554ad063536b59bb38bf1c/gadgets/top_file/gadget.yaml#L3-L6
- https://github.com/inspektor-gadget/inspektor-gadget/blob/0ac7b3d565b6f5775cccbe3595c1eea5a6b91871/gadgets/snapshot_process/gadget.yaml#L3-L5

### `GADGET_` macros

There are some `GADGET_` macros defined in in [include/gadget/macros.h](https://github.com/inspektor-gadget/inspektor-gadget/blob/7cdf64eac0d1f943152b602abee2421e961f355e/include/gadget/macros.h) that allow to set this information directly in the ebpf code of the gadget: 

- https://github.com/inspektor-gadget/inspektor-gadget/blob/9949c29dfe45e9622d8e5f5e603c954ea25294e2/gadgets/trace_exec/program.bpf.c#L60
- https://github.com/inspektor-gadget/inspektor-gadget/blob/79eb65512419557898554ad063536b59bb38bf1c/gadgets/top_file/program.bpf.c#L68
- https://github.com/inspektor-gadget/inspektor-gadget/blob/f01ffa9821808fb7b40673c873110f83dc0de2f6/gadgets/snapshot_process/program.bpf.c#L32

The current implementation uses the metadata file as the single source of truth, these macros are used  to populate the metadata if this is missing when running the gadget

https://github.com/inspektor-gadget/inspektor-gadget/blob/78058fb246bedc6107fb2660ac5b8e063e61c90a/pkg/gadgets/run/tracer/run.go#L166-L169

Or to generate it when --update-metadata is used when building the gadget image

https://github.com/inspektor-gadget/inspektor-gadget/blob/c067ec3411ce5b341bda9a3443c7e4f5cb39aeed/pkg/oci/build_metadata.go#L117-L119

## Issue 

While implementing #2132 @blanquicet mentioned this approach wasn't too clear and opened #2518, #2498. There is some discussion on #2518 and also on our developer sync, but there isn't an agreement yet. Additionally, #2512 appears to be changing this behavior too, so it's time to discuss and make a decision on this.

There are different ways to implement this:

### 1. Optional `GADGET_` macros and metadata 

(current approach)

As mentioned above, the current implementation allows to use any of the two ways. 

- Advantages
  - There's is more flexibility as users can choose the way better fits their needs
- Disadvantages
  - Having two ways to define the same creates confusion
  - We need to validate information matches when both ways are used at the same time

### 2. Mandatory `GADGET_` macros 

(Suggested in https://github.com/inspektor-gadget/inspektor-gadget/pull/2518#issuecomment-1956593517 and used by #2512)

- Advantages
  - The metadata file can always be generated from the code when using `--update-metadata`
  - There is not confusion as the macros are always needed
- Disadvantages
  - All users will be forced to use it

### 3. Mandatory metadata

- Advantages
  - Single source of truth 
- Disadvantages
  - Metadata will always be required, even if the user doesn't want to configure advance attributes
  - Users will need to write their metadata file by hand 

### 4. Best effort approach to automatically detect maps and structures used

We could have some logic to automatically detect the maps and structures used by the gadget.
- tracers: look for perf event array or ring buf and `event` structure
- toppers: look for hashmaps (value information can be automatically gathered from BTF information of the map)
- snapshotters: look for iterator programs and `entry` structure

- Advantages
  - Very simple as the user don't need to worry about specifying it
- Disadvantages
  - Whole approach is based on "magic" that relies on type names 
  - Could fail to automatically detect the information (not all hashmaps are used for sending "top" information)
  - Doesn't work if gadget implements multiple way to show information (multiple tracers, multiple toppers, etc.)

### PR DESCRIPTION
**Pull request by mauriciovasquezbernal**
This PR cleans up the logic to update and validate the metadata file.

It's mostly a cleanup after #2512 based  on the discussion of #2560 and the changes on #3149 

Fixes #2560

--- 

We're thinking different ways to generate the metadata for a gadget, perhaps we can use GetGadgetInfo(), for this reason this PR only updates the logic to follow the changes on #3149 but doesn't introduce tests as it could be changed very soon again.

### CODE REVIEW

**Comment by mauriciovasquezbernal** [comment]
Before going on the full motivation of the new approach, let me provide a bit of history on the different steps we've taken. The metadata file was initially conceived to fully describe a gadget, it contained which tracers, snapshotters, toppers, structures and parameters were exposed by it. It was the source of truth and the implementation fully relied on the information contained on this file to work. Then #2512 changed it by relying more on the information contained in the BPF object, this leaded to discuss the role of the metadata file in #2560, where we agreed to make the `GADGET_` macros mandatory. Then #3149 happened, making many parts (tracers, structures) of the metadata file useless. 

Now, the motivation for having this new approach (where we rely less on the metadata file is). This is mostly taken from #2560 plus some additional thoughts I have:
- There were two sources of information: (1) the ebpf object and (2) the metadata file. There was some complex logic to validate that both of them were aligned, also the fallback mechanisms when the metadata file was not present were confusing.
- Having two ways to define the gadget (`GADGET_` macros vs metadata) created confusion. 
- Forcing the user to create a metadata file adds more complexity to the gadget creation process. This file should be optional and only provide the "sugar" to make the gadget look better. For instance, if the user only wants to get json output, all column formatting options aren't needed.

File: pkg/gadgets/run/types/metadata.go [hunk 1]
```diff
@@ -24,9 +24,6 @@ import (
 	"github.com/hashicorp/go-multierror"
 	log "github.com/sirupsen/logrus"
 
-	"github.com/inspektor-gadget/inspektor-gadget/pkg/btfhelpers"
```
Comment (blanquicet) [inline pkg/gadgets/run/types/metadata.go:27]: After these changes and a couple of changes in the metadata files, I think can revert https://github.com/inspektor-gadget/inspektor-gadget/commit/eb6de7b7a77c7353d9c24fca1cfc05d255deee23. Do you want to do it in another PR?

**Review by blanquicet — APPROVED** [review]
LGTM! Thanks for this cleanup!

File: docs/devel/hello-world-gadget.md [hunk 2]
```diff
@@ -311,31 +311,18 @@ description: 'TODO: Fill the gadget description'
...
-        width: 16
-        alignment: left
-        ellipsis: end
-    - name: filename
-      description: 'TODO: Fill field description'
-      attributes:
-        width: 16
-        alignment: left
-        ellipsis: end
+      comm:
```
[the remark this answers is not in the document]
Comment (mauriciovasquezbernal) [reply docs/devel/hello-world-gadget.md:317]: Right, it's coming from the metadata package. I also don't know if we should populate it with all defaults.

File: pkg/gadgets/run/types/metadata.go (hunk 1 above)
Comment (mauriciovasquezbernal) [reply pkg/gadgets/run/types/metadata.go:27]: I'd prefer to wait a bit until we understand how to implement it again.

File: docs/devel/hello-world-gadget.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (blanquicet) [reply docs/devel/hello-world-gadget.md:310]: We either do that or provide a scheme where people can see the default values we are using. Otherwise, users will see columns displayed with a certain format and they'll not understand where such a configuration is coming from.

File: pkg/gadgets/run/types/metadata.go (hunk 1 above)
Comment (blanquicet) [reply pkg/gadgets/run/types/metadata.go:27]: I saw you created the function to populate the datasource yaml leaf. I tested it with `--update-metadata` and it worked. I ensures now the alphabetical order so some yaml files need to be updated, but other than that, I don't see what is missing. What am I not considering?

File: docs/devel/hello-world-gadget.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (mauriciovasquezbernal) [reply docs/devel/hello-world-gadget.md:310]: We can document the default annotations we use. The idea is to keep the metadata files clean without duplicating all default values.

File: pkg/gadgets/run/types/metadata.go (hunk 1 above)
Comment (mauriciovasquezbernal) [reply pkg/gadgets/run/types/metadata.go:27]: > I ensures now the alphabetical order so some yaml files need to be updated

This is what I was afraid of. Are you sure it's always the alphabetical order? If it's the case, we can re-enable that.

Comment (mauriciovasquezbernal) [reply pkg/gadgets/run/types/metadata.go:27]: I'm merging it right now to unblock other changes. I'll think about this later on.
